### ISSUE TITLE
why do callbacks catch all exceptions?

### ISSUE THREAD

**Issue by jon-deng** [matthias-k/cyipopt#72]
Hi,

Thanks for making this nice interface to ipopt.

I was wondering why in the callbacks to the objective, constraints, jacobian, etc. the functions will catch all exceptions and continue running? For example, in the code for the objective callback:
```python
cdef Bool objective_cb(
            Index n,
            Number* x,
            Bool new_x,
            Number* obj_value,
            UserDataPtr user_data
            ):

    log('objective_cb', logging.INFO)

    cdef object self = <object>user_data
    cdef Index i
    cdef np.ndarray[DTYPEd_t, ndim=1] _x = np.zeros((n,), dtype=DTYPEd)
    for i in range(n):
        _x[i] = x[i]
    try:
        obj_value[0] = self.__objective(_x)
    except:
        self.__exception = sys.exc_info()
    return True
```
all exceptions are caught.

I found that this can lead to really confusing behaviour. For example, if you make a small mistake in the objective function like opening a file that doesn't exist, an exception is thrown which can prevent additional stuff from computing. Since all exceptions are caught, excecution continues and ipopt will try to evaluate the gradient after. Since the objective function didn't compute properly this can lead to a new exception generated in the gradient that overwrites the previous exception. Eventually, ipopt reaches the convergence callback function and exits because there was an exception, but it's now the exception from the gradient and not the objective which makes it look the problem occurred in the gradient function, which is very confusing.

Is there a reason for catching the exceptions this way or is it possible to change it?

**Comment by moorepants** [comment]
It doesn't sound like a good design and the original designers are no longer involved. We welcome any pull requests to improve this.

**Comment by jon-deng** [comment]
I tried looking into getting around the exception catching but I think I figured why it was done the current way. C/Cython doesn't support exceptions in C functions so if an exception is raised in the callback there's no straightforward way to stop the c-code execution and pass the exception back to the python code that called ipopt. I think the current way of storing the exception and exiting at the intermediate callback is probably the best workaround then. Thanks for the quick reply.
